util/log: TestSecondaryGC failed

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1781212&tab=buildLog) on [master@120d53fc17f1f027c6075a2050f2991e369c5293](https://github.com/cockroachdb/cockroach/commits/120d53fc17f1f027c6075a2050f2991e369c5293):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC530034002
--- FAIL: TestSecondaryGC (0.09s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC530034002
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC530034002/gctest859008137", expected "/tmp/logTestSecondaryGC530034002"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1781602&tab=buildLog) on [master@ada086ebde92b35de12d66823d23faea85808606](https://github.com/cockroachdb/cockroach/commits/ada086ebde92b35de12d66823d23faea85808606):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC236528503
--- FAIL: TestSecondaryGC (0.07s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC236528503
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC236528503/gctest853237866", expected "/tmp/logTestSecondaryGC236528503"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1781727&tab=buildLog) on [master@aeb41dc98115da5d160e1cc01d2fa9c9e9c14c01](https://github.com/cockroachdb/cockroach/commits/aeb41dc98115da5d160e1cc01d2fa9c9e9c14c01):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC979885163
--- FAIL: TestSecondaryGC (0.12s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC979885163
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC979885163/gctest138420942", expected "/tmp/logTestSecondaryGC979885163"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1781742&tab=buildLog) on [master@ff2b6055b114929fb424bfe7948243d1be126d70](https://github.com/cockroachdb/cockroach/commits/ff2b6055b114929fb424bfe7948243d1be126d70):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC170573134
--- FAIL: TestSecondaryGC (0.11s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC170573134
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC170573134/gctest194741333", expected "/tmp/logTestSecondaryGC170573134"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1782044&tab=buildLog) on [master@064c4ea34b33b26894cbf1fde3cd31135e247455](https://github.com/cockroachdb/cockroach/commits/064c4ea34b33b26894cbf1fde3cd31135e247455):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC320993062
--- FAIL: TestSecondaryGC (0.07s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC320993062
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC320993062/gctest763840077", expected "/tmp/logTestSecondaryGC320993062"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1782263&tab=buildLog) on [master@064c4ea34b33b26894cbf1fde3cd31135e247455](https://github.com/cockroachdb/cockroach/commits/064c4ea34b33b26894cbf1fde3cd31135e247455):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC377944342
--- FAIL: TestSecondaryGC (0.08s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC377944342
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC377944342/gctest449058685", expected "/tmp/logTestSecondaryGC377944342"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1783171&tab=buildLog) on [master@d6a52a8f6f07b3b6b32af2243075e5365fb21c45](https://github.com/cockroachdb/cockroach/commits/d6a52a8f6f07b3b6b32af2243075e5365fb21c45):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC656847317
--- FAIL: TestSecondaryGC (0.07s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC656847317
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC656847317/gctest242996528", expected "/tmp/logTestSecondaryGC656847317"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1784634&tab=buildLog) on [master@d6a52a8f6f07b3b6b32af2243075e5365fb21c45](https://github.com/cockroachdb/cockroach/commits/d6a52a8f6f07b3b6b32af2243075e5365fb21c45):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC159022657
--- FAIL: TestSecondaryGC (0.08s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC159022657
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC159022657/gctest036595884", expected "/tmp/logTestSecondaryGC159022657"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1783189&tab=buildLog) on [master@d6a52a8f6f07b3b6b32af2243075e5365fb21c45](https://github.com/cockroachdb/cockroach/commits/d6a52a8f6f07b3b6b32af2243075e5365fb21c45):

```
=== RUN   TestSecondaryGC
test logs left over in: /tmp/logTestSecondaryGC467397039
--- FAIL: TestSecondaryGC (0.09s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestSecondaryGC467397039
    test_log_scope.go:137: unexpected logDir setting: set to "/tmp/logTestSecondaryGC467397039/gctest399280962", expected "/tmp/logTestSecondaryGC467397039"

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/cli/demo.go
+++ b/pkg/cli/demo.go
@@ -789,6 +789,10 @@ func runDemo(cmd *cobra.Command, gen workload.Generator) (err error) {
 
 	ctx := context.Background()
 
+	if err := checkTzDatabaseAvailability(ctx); err != nil {
+		return err
+	}
+
 	c, err := setupTransientCluster(ctx, cmd, gen)
 	defer c.cleanup()
 	if err != nil {
@@ -834,8 +838,6 @@ func runDemo(cmd *cobra.Command, gen workload.Generator) (err error) {
 #
 `, c.s.AdminURL())
 	}
-
-	checkTzDatabaseAvailability(ctx)
 
 	conn := makeSQLConn(c.connURL)
 	defer conn.Close()

--- a/pkg/cli/start.go
+++ b/pkg/cli/start.go
@@ -41,6 +41,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/engine"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
+	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
@@ -52,10 +53,10 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/sysutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble"
 	"github.com/cockroachdb/pebble/vfs"
 	opentracing "github.com/opentracing/opentracing-go"
-	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 	"google.golang.org/grpc"
 )
@@ -583,7 +584,9 @@ func runStart(cmd *cobra.Command, args []string, disableReplication bool) error 
 	// Until/unless CockroachDB embeds its own tz database, we want
 	// an early sanity check. It's better to inform the user early
 	// than to get surprising errors during SQL queries.
-	checkTzDatabaseAvailability(ctx)
+	if err := checkTzDatabaseAvailability(ctx); err != nil {
+		return errors.Wrap(err, "failed to initialize node")
+	}
 
 	// ReadyFn will be called when the server has started listening on
 	// its network sockets, but perhaps before it has done bootstrapping
@@ -1041,13 +1044,33 @@ func clientFlagsRPC() string {
 	return strings.Join(flags, " ")
 }
 
-func checkTzDatabaseAvailability(ctx context.Context) {
+func checkTzDatabaseAvailability(ctx context.Context) error {
 	if _, err := timeutil.LoadLocation("America/New_York"); err != nil {
-		log.Shout(ctx, log.Severity_ERROR,
-			"unable to load named time zones, time zone support will be degraded.\n"+
-				"Hint: check that the time zone database is installed on your system, or\n"+
+		log.Errorf(ctx, "timeutil.LoadLocation: %v", err)
+		reportedErr := errors.WithHint(
+			errors.WithIssueLink(
+				errors.New("unable to load named timezones"),
+				errors.IssueLink{IssueURL: unimplemented.MakeURL(36864)}),
+			"Check that the time zone database is installed on your system, or\n"+
 				"set the ZONEINFO environment variable to a Go time zone .zip archive.")
+
+		if envutil.EnvOrDefaultBool("COCKROACH_INCONSISTENT_TIME_ZONES", false) {
+			// The user tells us they really know what they want.
+			reportedErr := &formattedError{err: reportedErr}
+			log.Shout(ctx, log.Severity_WARNING, reportedErr.Error())
+		} else {
+			// Prevent a successful start.
+			//
+			// In the past, we were simply using log.Shout to emit an error,
+			// informing the user that startup could continue with degraded
+			// behavior.  However, usage demonstrated that users typically do
+			// not see the error and instead run into silently incorrect SQL
+			// results. To avoid this situation altogether, it's better to
+			// stop early.
+			return reportedErr
+		}
 	}
+	return nil
 }
 
 func reportConfiguration(ctx context.Context) {

--- a/pkg/cli/start.go
+++ b/pkg/cli/start.go
@@ -1124,7 +1124,7 @@ func logOutputDirectory() string {
 // occurs here and not in an OnInitialize function.
 func setupAndInitializeLoggingAndProfiling(
 	ctx context.Context, cmd *cobra.Command,
-) (*stop.Stopper, error) {
+) (stopper *stop.Stopper, err error) {
 	// Default the log directory to the "logs" subdirectory of the first
 	// non-memory store. If more than one non-memory stores is detected,
 	// print a warning.
@@ -1183,9 +1183,15 @@ func setupAndInitializeLoggingAndProfiling(
 		// directory too large.
 		log.StartGCDaemon(ctx)
 
-		// We have a valid logging directory. Configure Pebble/RocksDB to log into it.
-		engine.InitPebbleLogger(ctx)
-		engine.InitRocksDBLogger(ctx)
+		defer func() {
+			if stopper != nil {
+				// When the function complete successfully, start the loggers
+				// for the storage engines. We need to do this at the end
+				// because we need to register the loggers.
+				stopper.AddCloser(engine.InitPebbleLogger(ctx))
+				stopper.AddCloser(engine.InitRocksDBLogger(ctx))
+			}
+		}()
 	}
 
 	outputDirectory := "."
@@ -1241,7 +1247,7 @@ func setupAndInitializeLoggingAndProfiling(
 	// Disable Stopper task tracking as performing that call site tracking is
 	// moderately expensive (certainly outweighing the infrequent benefit it
 	// provides).
-	stopper := initBacktrace(outputDirectory)
+	stopper = initBacktrace(outputDirectory)
 	log.Event(ctx, "initialized profiles")
 
 	return stopper, nil

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -772,6 +772,9 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 			internalExecutor,
 		),
 
+		// Note: don't forget to add the secondary loggers as closers
+		// on the Stopper, below.
+
 		ExecLogger: log.NewSecondaryLogger(
 			loggerCtx, nil /* dirName */, "sql-exec",
 			true /* enableGc */, false /*forceSyncWrites*/, true, /* enableMsgCount */
@@ -790,6 +793,10 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 		QueryCache:                 querycache.New(s.cfg.SQLQueryCacheSize),
 		ProtectedTimestampProvider: s.protectedtsProvider,
 	}
+
+	s.stopper.AddCloser(execCfg.ExecLogger)
+	s.stopper.AddCloser(execCfg.AuditLogger)
+	s.stopper.AddCloser(execCfg.SlowQueryLogger)
 
 	if sqlSchemaChangerTestingKnobs := s.cfg.TestingKnobs.SQLSchemaChanger; sqlSchemaChangerTestingKnobs != nil {
 		execCfg.SchemaChangerTestingKnobs = sqlSchemaChangerTestingKnobs.(*sql.SchemaChangerTestingKnobs)

--- a/pkg/storage/engine/pebble.go
+++ b/pkg/storage/engine/pebble.go
@@ -329,10 +329,12 @@ var pebbleLog *log.SecondaryLogger
 
 // InitPebbleLogger initializes the logger to use for Pebble log messages. If
 // not called, WARNING, ERROR, and FATAL logs will be output to the normal
-// CockroachDB log.
-func InitPebbleLogger(ctx context.Context) {
+// CockroachDB log. The caller is responsible for ensuring the
+// Close() method is eventually called on the new logger.
+func InitPebbleLogger(ctx context.Context) *log.SecondaryLogger {
 	pebbleLog = log.NewSecondaryLogger(ctx, nil, "pebble",
 		true /* enableGC */, false /* forceSyncWrites */, false /* enableMsgCount */)
+	return pebbleLog
 }
 
 type pebbleLogger struct {

--- a/pkg/storage/engine/rocksdb.go
+++ b/pkg/storage/engine/rocksdb.go
@@ -84,10 +84,12 @@ var rocksdbLogger *log.SecondaryLogger
 
 // InitRocksDBLogger initializes the logger to use for RocksDB log messages. If
 // not called, WARNING, ERROR, and FATAL logs will be output to the normal
-// CockroachDB log.
-func InitRocksDBLogger(ctx context.Context) {
+// CockroachDB log. The caller is responsible for ensuring the
+// Close() method is eventually called on the new logger.
+func InitRocksDBLogger(ctx context.Context) *log.SecondaryLogger {
 	rocksdbLogger = log.NewSecondaryLogger(ctx, nil, "rocksdb",
 		true /* enableGC */, false /* forceSyncWrites */, false /* enableMsgCount */)
+	return rocksdbLogger
 }
 
 //export rocksDBLog

--- a/pkg/util/errorutil/unimplemented/unimplemented.go
+++ b/pkg/util/errorutil/unimplemented/unimplemented.go
@@ -77,7 +77,7 @@ func unimplementedInternal(
 	// Create the issue link.
 	link := errors.IssueLink{Detail: detail}
 	if issue > 0 {
-		link.IssueURL = makeURL(issue)
+		link.IssueURL = MakeURL(issue)
 	}
 
 	// Instantiate the base error.
@@ -108,6 +108,7 @@ func unimplementedInternal(
 	return err
 }
 
-func makeURL(issue int) string {
+// MakeURL produces a URL to a CockroachDB issue.
+func MakeURL(issue int) string {
 	return fmt.Sprintf("https://github.com/cockroachdb/cockroach/issues/%d", issue)
 }

--- a/pkg/util/errorutil/unimplemented/unimplemented_test.go
+++ b/pkg/util/errorutil/unimplemented/unimplemented_test.go
@@ -50,7 +50,7 @@ func TestUnimplemented(t *testing.T) {
 				}
 				if test.expIssue != 0 {
 					ref := fmt.Sprintf("%s\nSee: %s",
-						errors.UnimplementedErrorHint, makeURL(test.expIssue))
+						errors.UnimplementedErrorHint, MakeURL(test.expIssue))
 					if hint == ref {
 						found |= 2
 					}
@@ -79,7 +79,7 @@ func TestUnimplemented(t *testing.T) {
 				}
 
 				if test.expIssue != 0 {
-					url := makeURL(test.expIssue)
+					url := MakeURL(test.expIssue)
 					if links[0].IssueURL != url {
 						t.Errorf("expected link url %q, got %q", url, links[0].IssueURL)
 					}

--- a/pkg/util/log/log_gc_test.go
+++ b/pkg/util/log/log_gc_test.go
@@ -54,6 +54,7 @@ func TestSecondaryGC(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 	l := NewSecondaryLogger(ctx, &tmpDirName, "woo", false /*enableGc*/, false /*syncWrites*/, true /*msgCount*/)
+	defer l.Close()
 
 	testLogGC(t, &l.logger, l.Logf)
 }

--- a/pkg/util/log/secondary_log.go
+++ b/pkg/util/log/secondary_log.go
@@ -44,6 +44,9 @@ var secondaryLogRegistry struct {
 // in which case it specifies the directory for that new logger.
 //
 // The logger's GC daemon stops when the provided context is canceled.
+//
+// The caller is responsible for ensuring the Close() method is
+// eventually called.
 func NewSecondaryLogger(
 	ctx context.Context,
 	dirName *DirName,
@@ -79,31 +82,28 @@ func NewSecondaryLogger(
 	secondaryLogRegistry.mu.loggers = append(secondaryLogRegistry.mu.loggers, l)
 	secondaryLogRegistry.mu.Unlock()
 
-	// Make the registry forget about this logger when the context is
-	// canceled. This avoids stacking many secondary loggers together
-	// when there are subsequent tests starting servers in the same
-	// package.
-	go func() {
-		// Wait for the stopper to say "goodbye".
-		<-ctx.Done()
-		// De-register.
-		secondaryLogRegistry.mu.Lock()
-		defer secondaryLogRegistry.mu.Unlock()
-		for i, thatLogger := range secondaryLogRegistry.mu.loggers {
-			if thatLogger != l {
-				continue
-			}
-			secondaryLogRegistry.mu.loggers = append(secondaryLogRegistry.mu.loggers[:i], secondaryLogRegistry.mu.loggers[i+1:]...)
-			return
-		}
-	}()
-
 	if enableGc {
 		// Start the log file GC for the secondary logger.
 		go l.logger.gcDaemon(ctx)
 	}
 
 	return l
+}
+
+// Close implements the stopper.Closer interface.
+func (l *SecondaryLogger) Close() {
+	// Make the registry forget about this logger. This avoids
+	// stacking many secondary loggers together when there are
+	// subsequent tests starting servers in the same package.
+	secondaryLogRegistry.mu.Lock()
+	defer secondaryLogRegistry.mu.Unlock()
+	for i, thatLogger := range secondaryLogRegistry.mu.loggers {
+		if thatLogger != l {
+			continue
+		}
+		secondaryLogRegistry.mu.loggers = append(secondaryLogRegistry.mu.loggers[:i], secondaryLogRegistry.mu.loggers[i+1:]...)
+		return
+	}
 }
 
 func (l *SecondaryLogger) output(

--- a/pkg/util/log/secondary_log_test.go
+++ b/pkg/util/log/secondary_log_test.go
@@ -35,6 +35,7 @@ func TestSecondaryLog(t *testing.T) {
 
 	// Make a new logger, in the same directory.
 	l := NewSecondaryLogger(ctx, &mainLog.logDir, "woo", true, false, true)
+	defer l.Close()
 
 	// Interleave some messages.
 	Infof(context.Background(), "test1")
@@ -88,6 +89,7 @@ func TestRedirectStderrWithSecondaryLoggersActive(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 	l := NewSecondaryLogger(ctx, &mainLog.logDir, "woo", true, false, true)
+	defer l.Close()
 
 	// Log something on the secondary logger.
 	l.Logf(context.Background(), "test456")
@@ -125,6 +127,7 @@ func TestListLogFilesIncludeSecondaryLogs(t *testing.T) {
 
 	// Make a new logger, in the same directory.
 	l := NewSecondaryLogger(ctx, &mainLog.logDir, "woo", true, false, true)
+	defer l.Close()
 
 	// Emit some logging and ensure the files gets created.
 	l.Logf(ctx, "story time")

